### PR TITLE
sort: reorganize header, add overloads for custom comparator

### PR DESCRIPTION
**Pull request by fnrizzi**
# Content

This PR is doing 2 things: 
(A) reorganize the sorting header files to improve readability, granularity and clearly separate what is public from impl
(b) adds 2 new overloads to `Kokkos::sort` to support a custom comparator (behavior of overloads in develop is unchanged) 
Disclaimer: these two things are done in the same PR because they are closely related 

### Before: 
```
algorithms/
├── CMakeLists.txt
├── src
│   ├── CMakeLists.txt
│   ├── KokkosAlgorithms_dummy.cpp
│   ├── Kokkos_NestedSort.hpp
│   ├── Kokkos_Random.hpp
│   ├── Kokkos_Sort.hpp
│   ├── Kokkos_StdAlgorithms.hpp
```
Issues with this: `Kokkos_Sort.hpp` contains *all* the public APIs and *impl* for sort and bin sort, making it hard to follow and understand what is going on, and exposes all the impl details too. The same is happening for `Kokkos_NestedSort.hpp`

### This PR addresses the above as follows
```
algorithms/
├── CMakeLists.txt
├── src
│   ├── CMakeLists.txt
│   ├── KokkosAlgorithms_dummy.cpp
│   ├── Kokkos_Random.hpp
│   ├── Kokkos_Sort.hpp
│   ├── Kokkos_StdAlgorithms.hpp
│   ├── sorting
│   │   ├── Kokkos_BinOps.hpp
│   │   ├── Kokkos_BinSort.hpp
│   │   ├── Kokkos_NestedSort.hpp
│   │   ├── Kokkos_Sort.hpp
│   │   └── impl
│   │       ├── Kokkos_BinSort.hpp
│   │       ├── Kokkos_NestedSort.hpp
│   │       └── Kokkos_SortFreeFunction.hpp
```
- Public API and impl details are separated, therefore not exposing impl details to users
- the public API is simplified and cleaned a bit (less layers jump from API to impl) to make it easier to follow which impl is used where 
- `Kokkos_Sort.hpp` remains the header to include to access *all* sorting functionalities. 

<br>

### (B1) New overloads added to `Kokkos::sort` 

```cpp
template <class ExecutionSpace, class CompType, class DataType, class... Properties>
void sort([[maybe_unused]] const ExecutionSpace& exec,
          const Kokkos::View<DataType, Properties...>& view,
          CompType const& comp) 
{
   /*
    accepts rank-1 view with layout{left,right,stride}

    if hostaccesible: 
        calls std::sort
    if cuda: 
        calls thrust::sort
    if sycl:
        if layoutleft or right, calls oneapi
        if layoutstride, copies to host, calls std::sort and copies back
    else:
        copies view to host, calls std::sort and copies back
   */
}

template <class CompType, class DataType, class... Properties>
void sort(const Kokkos::View<DataType, Properties...>& view, CompType const& comp){
  // calls overload above using the view execution space
}
```

- where the fallback case was based on a discussion with @crtrott  and @dalg24. Note that this is potentially the place where we can later optimize for example implementing our own quicksort or what not.

### (B2) Behavior of overloads currently in develop remains unchanged 

```cpp
template <class DataType, class... Properties>
void sort(const executionspace &, const Kokkos::View<DataType, Properties...>& view){          // (1) 
  /* accepts regular view, rank-1

    if hostaccesible: calls std sort
    if cuda:          calls thrust
    if sycl:          must be layoutleft or right, calls oneapi
    else: calls binsort
  */
}

template <class ViewType> 
void sort(ViewType const& view){                                                                // (2) 
  // calls overload (1) above using the view execution space
}

template <class ExecutionSpace, class ViewType>
void sort(const ExecutionSpace& exec, ViewType view, size_t const begin, size_t const end){     // (3) 
  // accepts regular view or dynamic view, rank1 or rank2
  // calls binsort
}

template <class ViewType>
void sort(ViewType view, size_t const begin, size_t const end){                                 // (4) 
  // calls overload (3) above using the view execution space
}
```

<br>

# Notes and related things 

- This PR also sets things up such that later it will be easier to incorporate #5507 once ready, for example as a specialization for sorting view/ranges with "numbers" 

- https://github.com/kokkos/kokkos/issues/6216

### CODE REVIEW

**Comment by fnrizzi** [comment]
retest this please

File: algorithms/src/sorting/Kokkos_Sort.hpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+		"View argument!");
+
+  // trivial cases
+  if (view.extent(0) <= 1) { return; }
+
+  // non trivial
+  if constexpr (SpaceAccessibility<HostSpace, MemSpace>::accessible) {
+    auto first = ::Kokkos::Experimental::begin(view);
+    auto last  = ::Kokkos::Experimental::end(view);
+    std::sort(first, last);
```
Comment (brian-kelley) [inline algorithms/src/sorting/Kokkos_Sort.hpp:53]: Should std::sort be used only when ExecutionSpace == Serial? ExecutionSpace could be host parallel but std::sort with no execution policy will run sequentially, AFAIK

File: algorithms/src/sorting/Kokkos_Sort.hpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	   std::is_same_v<typename ViewType::array_layout, LayoutStride>),
+      "Kokkos::sort: supports 1D Views with LayoutRight, LayoutLeft or LayoutStride.");
+
+  // trivial cases
+  if (view.extent(0) <= 1) { return; }
+
+  if constexpr (SpaceAccessibility<HostSpace, MemSpace>::accessible) {
+    auto first = ::Kokkos::Experimental::begin(view);
+    auto last  = ::Kokkos::Experimental::end(view);
+    std::sort(first, last, comp);
```
Comment (brian-kelley) [inline algorithms/src/sorting/Kokkos_Sort.hpp:106]: Same as above

File: algorithms/src/sorting/impl/Kokkos_SortFreeFunction.hpp [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+#include <oneapi/dpl/algorithm>
+#endif
+
+namespace Kokkos {
+namespace Impl {
+// clang-format off
+
+#if defined(KOKKOS_ENABLE_CUDA)
+template <class DataType, class... Properties, class... ComparatorOrEmpty>
+void sort_cudathurst_and_fence_exec(const Cuda& exec,
```
Comment (brian-kelley) [inline algorithms/src/sorting/impl/Kokkos_SortFreeFunction.hpp:76]:
```suggestion
void sort_cudathrust_and_fence_exec(const Cuda& exec,
```

File: algorithms/src/sorting/impl/Kokkos_SortFreeFunction.hpp [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+// specialize cases for sorting without comparator
+//
+// --------------------------------------------------
+
+#if defined(KOKKOS_ENABLE_CUDA)
+template <class DataType, class... Properties>
+void sort_device_view_without_comparator_and_fence_exec(
+    const Cuda& exec, const Kokkos::View<DataType, Properties...>& view)
+{
+  sort_cudathurst_and_fence_exec(exec, view);
```
Comment (brian-kelley) [inline algorithms/src/sorting/impl/Kokkos_SortFreeFunction.hpp:242]:
```suggestion
  sort_cudathrust_and_fence_exec(exec, view);
```

**Review by brian-kelley — CHANGES_REQUESTED** [review]
I like the file reorganization, thanks!

File: algorithms/src/sorting/impl/Kokkos_SortFreeFunction.hpp (hunk 3 above)
Comment (fnrizzi) [reply algorithms/src/sorting/impl/Kokkos_SortFreeFunction.hpp]: good catch!

File: algorithms/unit_tests/TestSortCustomComp.hpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+#include <Kokkos_Sort.hpp>
+#include <TestStdAlgorithmsCommon.hpp>
+
+namespace Test {
+namespace SortWithComp {
+
+template <class T>
+struct MyComp {
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(T a, T b) const { return a < b; }
```
[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/unit_tests/TestSortCustomComp.hpp]: done!

File: algorithms/unit_tests/TestSortCustomComp.hpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    auto dataView_h = stdalgos::create_host_space_copy(dataView);
+    stdalgos::compare_views(dataViewBeforeOp_h, dataView_h);
+
+    // To actually check that Kokkos::sort used the custom
+    // comparator MyComp, we should have a result in non-ascending order.
+    // We can verify this by running std::is_sorted and if that returns
+    // false, then it means everything ran as expected
+    namespace KE = Kokkos::Experimental;
+    ASSERT_FALSE(std::is_sorted( KE::cbegin(dataView_h), KE::cend(dataView_h)));
```
Comment (brian-kelley) [inline algorithms/unit_tests/TestSortCustomComp.hpp:116]:
```suggestion
```

**Review by brian-kelley — COMMENTED** [review]
Since ``compare_views()`` above will exactly check the correctness of each dataView element, I would just skip checking this. And for api == 2 or 3, dataView would be sorted ascending, right?

File: algorithms/unit_tests/TestSortCustomComp.hpp (hunk 6 above)
Comment (fnrizzi) [reply algorithms/unit_tests/TestSortCustomComp.hpp:116]: oh i forgot to commit the correct code , this is missing an if

**Review by brian-kelley — APPROVED** [review]

**Comment by fnrizzi** [comment]
CI error unrelated: NVC++-S-0000-Internal compiler error. extractor: bad opc

File: algorithms/unit_tests/TestSortCustomComp.hpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+#include <Kokkos_Random.hpp>
+#include <Kokkos_Sort.hpp>
+#include <TestStdAlgorithmsCommon.hpp>
+
+namespace Test {
+namespace SortWithComp {
+
+template <class T>
+struct MyComp {
+  KOKKOS_INLINE_FUNCTION
```
Comment (dalg24) [inline algorithms/unit_tests/TestSortCustomComp.hpp:31]:
```suggestion
  KOKKOS_FUNCTION
```

File: algorithms/unit_tests/TestSortCustomComp.hpp [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+TEST(TEST_CATEGORY, SortWithCustomComparator) {
+  using ExeSpace = TEST_EXECSPACE;
+
+#if defined(KOKKOS_ENABLE_SYCL)
+  for (int api = 0; api < 2; api++) {
+    run_all_scenarios<ExeSpace, stdalgos::DynamicTag, int>(api);
+    run_all_scenarios<ExeSpace, stdalgos::DynamicTag, double>(api);
+  }  // end for
```
Comment (dalg24) [inline algorithms/unit_tests/TestSortCustomComp.hpp:129]: Is that a FIXME_SYCL?

File: algorithms/src/sorting/impl/Kokkos_SortFreeFunction.hpp [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+
+namespace Kokkos {
+namespace Impl {
+// clang-format off
+
+#if defined(KOKKOS_ENABLE_CUDA)
+template <class DataType, class... Properties, class... ComparatorOrEmpty>
+void sort_cudathrust_and_fence_exec(const Cuda& exec,
+                     const Kokkos::View<DataType, Properties...>& view,
+                     ComparatorOrEmpty&&... compOrEmpty)
```
Comment (dalg24) [inline algorithms/src/sorting/impl/Kokkos_SortFreeFunction.hpp:78]: (optional) ComparatorOrEmpty -> MaybeComparator

File: algorithms/src/sorting/Kokkos_Sort.hpp [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+  } else {
+    Impl::sort_device_view_with_comparator_and_fence_exec(exec, view, comp);
```
Comment (dalg24) [inline algorithms/src/sorting/Kokkos_Sort.hpp:108]: Why "and fence"?  The overload that takes an execution space as 1st argument is supposed to be asynchronous.

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: algorithms/src/sorting/Kokkos_BinOps.hpp [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_BIN_OPS_HPP_
+#define KOKKOS_BIN_OPS_HPP_
+
+#include <Kokkos_Core.hpp>
```
[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/sorting/Kokkos_BinOps.hpp:20]: ok thank you will try that

File: algorithms/src/sorting/Kokkos_BinOps.hpp [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+namespace Kokkos {
+
+template <class KeyViewType>
+struct BinOp1D {
+  int max_bins_ = {};
+  double mul_   = {};
+  double min_   = {};
+
+  BinOp1D() = default;
```
[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/sorting/Kokkos_BinOps.hpp:30]: ops, not sure how did this end up being reverted. thank you for catching it! (i fixed also the BinOp3D)

File: algorithms/src/sorting/Kokkos_Sort.hpp [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+
+#include "impl/Kokkos_SortFreeFunction.hpp"
+#include <Kokkos_Core.hpp>
+#include <std_algorithms/Kokkos_BeginEnd.hpp>
+#include <std_algorithms/Kokkos_Copy.hpp>
+#include <Kokkos_DynamicView.hpp>
+#include <algorithm>
+
+namespace Kokkos {
+// clang-format off
```
[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/sorting/Kokkos_Sort.hpp:28]: because the indentation was really weird, making IMO the various overloads very hard to read/skim. 
I can let it format them if you really insist

[the remark this answers is not in the document]
Comment (dalg24) [reply algorithms/src/sorting/Kokkos_Sort.hpp:28]: I would prefer you restricting the disabling format precisely where appropriate.
The smaller the scope the better.

File: algorithms/src/sorting/Kokkos_Sort.hpp (hunk 10 above)
Comment (fnrizzi) [reply algorithms/src/sorting/Kokkos_Sort.hpp:108]: my bad, i was mislead by https://kokkos.github.io/kokkos-core-wiki/API/algorithms/Sort.html#additional-information

Comment (fnrizzi) [reply algorithms/src/sorting/Kokkos_Sort.hpp:108]: I will fix the fencing where needed

Comment (dalg24) [reply algorithms/src/sorting/Kokkos_Sort.hpp:108]: Yes

**Comment by fnrizzi** [comment]
> Please rebase on top of `develop` to get rid of the duplicated commits.

what duplicated commits?

**Comment by masterleinad** [comment]
> what duplicated commits?

This pull request shares a lot of commits with https://github.com/kokkos/kokkos/pull/6081/commits.

**Comment by fnrizzi** [comment]
will try to split this

**Comment by fnrizzi** [comment]
> https://github.com/kokkos/kokkos/pull/6081/commits

ah ok sorry i misunderstood!

**Comment by fnrizzi** [comment]
closing since this will be replaced by smaller PRs
